File: Leetcode/dfs_recursion_backtracking/77_Combinations/Solution.java
class Solution {
    public List<List<Integer>> combine(int n, int k) {
        int[] nums = new int[n];
        for(int i = 0; i<n; ++i){ nums[i]=i+1; }
        List<List<Integer>> res = new LinkedList<>();
        dfs(res, new LinkedList<Integer>(), nums, k, 0);
        return res;
    }
    private void dfs(List<List<Integer>> res, LinkedList<Integer> temp,
                    int[] nums, int k, int start){
        if(k==0){
            res.add(new LinkedList<Integer>(temp));
            return;
        }
        for(int i = start; i<nums.length; ++i){
            temp.add(nums[i]);
            dfs(res, temp, nums, k-1, i+1);
            temp.pollLast();
        }
    }
}

File: Leetcode/dfs_recursion_backtracking/77_Combinations/Solution.py
class Solution(object):
    def combine(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: List[List[int]]
        """
        nums = [i for i in range(1, n+1)]
        res = []
        def dfs(temp, k, start):
            if k == 0:
                res.append(list(temp))
                return
            for i in range(start, len(nums)):
                temp.append(nums[i])
                dfs(temp, k-1, i+1)
                temp.pop()
        dfs([], k, 0)
        return res
